x-pack/metricbeat/module/aws/awshealth/awshealth.go

```diff
@@ -0,0 +1,436 @@
+		return nil, err
+	}
+
+	return &MetricSet{
+		MetricSet: metricSet,
+		logger:    logger,
+		Config:    config.Config,
+	}, nil
+}
+
+// Fetch method implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(ctx context.Context, report mb.ReporterV2) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	var config aws.Config
+	err := m.Module().UnpackConfig(&config)
+	if err != nil {
+		return err
+	}
+
+	// Get startDate and endDate
+	// startDate, endDate := getStartDateEndDate(m.Period)
+
+	awsConfig := m.MetricSet.AwsConfig.Copy()
+
+	health_client := health.NewFromConfig(awsConfig, func(o *health.Options) {
+		if config.AWSConfig.FIPSEnabled {
+			o.EndpointOptions.UseFIPSEndpoint = awssdk.FIPSEndpointStateEnabled
+		}
+	})
+
+	events := m.getEventsSummary(ctx, health_client)
+	for _, event := range events {
+		report.Event(event)
+	}
+
+	return nil
+}
+
+// Make call to DescribeEvents()
+// Returns information about events that meet the specified filter criteria. Events are returned in a summary form and do not include the detailed description, any additional metadata that depends on the event type, or any affected resources.
+func (m *MetricSet) getEventsSummary(
+	ctx context.Context,
+	awsHealth *health.Client,
+) []mb.Event {
+	var events []mb.Event
+	eventFilter := types.EventFilter{
+		EventStatusCodes: []types.EventStatusCode{
+			types.EventStatusCodeUpcoming,
+			types.EventStatusCodeOpen,
+		},
+	}
+
+	var (
+		nextTokenString string
+		eventOutput     *health.DescribeEventsOutput
+		err             error
+		wg              sync.WaitGroup
+	)
+	errCh := make(chan error, maxResults)
+	c := make(chan HealthDetails, maxResults)
+
+	for {
+		// When invoking the DescribeEvents for the first time, there must not exist any NextToken.
+		// DescribeEvents API call will return the next token if there are more records left for querying
+		// If there exist no further records to fetch, next toke will be empty.
+		if nextTokenString == "" {
+			eventOutput, err = awsHealth.DescribeEvents(ctx,
+				&health.DescribeEventsInput{
+					Filter:     &eventFilter,
+					MaxResults: &maxResults,
+				},
+			)
+		} else {
+			eventOutput, err = awsHealth.DescribeEvents(ctx,
+				&health.DescribeEventsInput{
+					Filter:     &eventFilter,
+					MaxResults: &maxResults,
+					NextToken:  &nextTokenString,
+				},
+			)
+		}
+		if err != nil {
+			err = fmt.Errorf("[AWS Health] DescribeEvents failed with : %w", err)
+			m.Logger().Error(err.Error())
+			return nil
+		}
+		ets := eventOutput.Events
+		select {
+		case <-ctx.Done():
+			// Context cancelled, handle graceful termination
+			close(c)
+			return nil
+		default:
+			// Context not cancelled, proceed with the function
+		}
+
+		for _, et := range ets {
```

Comment:
Refer diff patch.


Micro optimization tip:
So for ... range create a copy of the indexed element whereas normal indexing does not. See: https://go.dev/tour/moretypes/16

Also as in Go everything is passed by value, the argument you are sending to the closure `go func(et types.Event)` again creates a copy of the event.

So, 2 copies. So, we can try to minimize that. See the diff patch for a less expensive alternate way. 

We are still binding a value to the closure because of the gotcha with goroutines and closure in Go. See:
https://blog.cloudflare.com/a-go-gotcha-when-closures-and-goroutines-collide

Reply:
https://github.com/elastic/beats/pull/38370#issuecomment-2011254877


As the diff patch is hidden in this PR due to the large number of build messages, I am pasting the link to it here.